File: lib/messages/messages.dart
// Message definitions for Dart side of rinf bridge.
// These are hand-written stubs — rinf codegen will replace them.
// Maps to rust/src/messages.rs

class OpenPdfRequest {
  final String filePath;
  OpenPdfRequest({required this.filePath});
}

class OcrPageRequest {
  final int pageNum;
  final String side;
  final int? psm;
  final String? language;
  OcrPageRequest({required this.pageNum, required this.side, this.psm, this.language});
}

class PdfOpenedSignal {
  final String filePath;
  final int totalPages;
  final int spreadsDetected;
  final List<PageInfo> pages;
  PdfOpenedSignal({required this.filePath, required this.totalPages, required this.spreadsDetected, required this.pages});
}

class PageInfo {
  final int pageNum;
  final int width;
  final int height;
  final bool isSpread;
  final double aspectRatio;
  PageInfo({required this.pageNum, required this.width, required this.height, required this.isSpread, required this.aspectRatio});
}

class OcrCompletedSignal {
  final int pageNum;
  final String side;
  final String text;
  final double confidence;
  final int wordCount;
  OcrCompletedSignal({required this.pageNum, required this.side, required this.text, required this.confidence, required this.wordCount});
}

class TextCleanedSignal {
  final String cleanedText;
  final int rawLength;
  final int cleanedLength;
  final int headersRemoved;
  TextCleanedSignal({required this.cleanedText, required this.rawLength, required this.cleanedLength, required this.headersRemoved});
}

class ProgressSignal {
  final int current;
  final int total;
  final String message;
  final double percentage;
  ProgressSignal({required this.current, required this.total, required this.message, required this.percentage});
}

class ErrorSignal {
  final String message;
  final String code;
  ErrorSignal({required this.message, required this.code});
}

File: rust/src/messages.rs
// Message definitions for Rust <-> Dart communication via rinf.
// These structs are serialized and sent across the FFI boundary.
use rinf::DartSignal;
use serde::{Deserialize, Serialize};

use crate::ocr_engine::OcrResult;
use crate::pdf_processor::PageInfo;
use crate::text_cleaner::CleanResult;

// ============ Dart -> Rust (requests) ============

/// Request to open and analyze a PDF
#[derive(Debug, Serialize, Deserialize)]
pub struct OpenPdfRequest {
    pub file_path: String,
}

/// Request to OCR a specific page
#[derive(Debug, Serialize, Deserialize)]
pub struct OcrPageRequest {
    pub page_num: u32,
    pub side: String, // "left", "right", "single"
    pub psm: Option<i32>,
    pub language: Option<String>,
}

/// Request to clean text with specified headers
#[derive(Debug, Serialize, Deserialize)]
pub struct CleanTextRequest {
    pub raw_text: String,
    pub headers: Vec<String>,
}

// ============ Rust -> Dart (responses/signals) ============

/// Signal: PDF opened successfully with page info
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PdfOpenedSignal {
    pub file_path: String,
    pub total_pages: u32,
    pub spreads_detected: u32,
    pub pages: Vec<PageInfo>,
}

/// Signal: OCR completed for a page
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OcrCompletedSignal {
    pub page_num: u32,
    pub side: String,
    pub text: String,
    pub confidence: f32,
    pub word_count: u32,
}

/// Signal: Text cleaned
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TextCleanedSignal {
    pub cleaned_text: String,
    pub raw_length: u32,
    pub cleaned_length: u32,
    pub headers_removed: u32,
}

/// Signal: Processing progress update
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProgressSignal {
    pub current: u32,
    pub total: u32,
    pub message: String,
    pub percentage: f32,
}

/// Signal: Error occurred
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorSignal {
    pub message: String,
    pub code: String,
}
